src/app/ReadClient.cpp

```diff
@@ -517,6 +541,118 @@ CHIP_ERROR ReadClient::ProcessAttributeReportIBs(TLV::TLVReader & aAttributeRepo
+{
+    CHIP_ERROR err               = CHIP_NO_ERROR;
+    uint64_t timeStampVal        = 0;
+    bool hasSystemTimestamp      = false;
+    bool hasEpochTimestamp       = false;
+    bool hasDeltaSystemTimestamp = false;
+    bool hasDeltaEpochTimestamp  = false;
+    err                          = aEventData.GetDeltaSystemTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mDeltaTimestamp.mType  = Timestamp::Type::kSystem;
+        aEventHeader.mDeltaTimestamp.mValue = timeStampVal;
+        hasDeltaSystemTimestamp             = true;
+    }
+
+    err = aEventData.GetDeltaEpochTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mDeltaTimestamp.mType  = Timestamp::Type::kEpoch;
+        aEventHeader.mDeltaTimestamp.mValue = timeStampVal;
+        hasDeltaEpochTimestamp              = true;
+    }
+
+    err = aEventData.GetSystemTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mTimestamp.mType  = Timestamp::Type::kSystem;
+        aEventHeader.mTimestamp.mValue = timeStampVal;
+        hasSystemTimestamp             = true;
+    }
+
+    err = aEventData.GetEpochTimestamp(&timeStampVal);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        aEventHeader.mDeltaTimestamp.mType  = Timestamp::Type::kEpoch;
+        aEventHeader.mDeltaTimestamp.mValue = timeStampVal;
+        hasEpochTimestamp                   = true;
+    }
+
+    if ((hasSystemTimestamp && !hasEpochTimestamp && !hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && hasEpochTimestamp && !hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && !hasEpochTimestamp && hasDeltaSystemTimestamp && !hasDeltaEpochTimestamp) ||
+        (!hasSystemTimestamp && !hasEpochTimestamp && !hasDeltaSystemTimestamp && hasDeltaEpochTimestamp))
+    {
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_IM_MALFORMED_EVENT_DATA_ELEMENT;
+}
+
+CHIP_ERROR ReadClient::ProcessEventReportIBs(TLV::TLVReader & aEventReportIBsReader)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    while (CHIP_NO_ERROR == (err = aEventReportIBsReader.Next()))
+    {
+        TLV::TLVReader dataReader;
+        EventReportIB::Parser report;
+        EventDataIB::Parser data;
+        EventStatusIB::Parser status;
+        EventPathIB::Parser path;
+        ClusterInfo clusterInfo;
+        StatusIB statusIB;
+        EventHeader header;
+        uint8_t priorityLevel = 0;
+
+        TLV::TLVReader reader = aEventReportIBsReader;
+        ReturnErrorOnFailure(report.Init(reader));
+        // EventStatus and EventData would coexist
```

Comment:
The spec stipulates that only one of the two exists.

Reply:
And for starters, I'd rather we not implement EventStatus (and add that later, since we don't actually have the plumbing on the server side to do this correctly).

Reply:
Sure, would adjust to either or

Reply:
removed.